src/libraries/Microsoft.PowerFx.Connectors/Execution/FormulaValueSerializer.cs

```diff
@@ -0,0 +1,188 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Linq;
+using Microsoft.OpenApi.Models;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Connectors.Execution
+{
+    internal abstract class FormulaValueSerializer
+    {
+        protected abstract string GetResult();
+
+        protected abstract void StartSerialization(OpenApiSchema schema);
+
+        protected abstract void StartObject(string name = null);
+
+        protected abstract void EndObject();
+
+        protected abstract void StartArray(string name = null);
+
+        protected abstract void StartArrayElement(string name);
+
+        protected abstract void EndArray();
+
+        protected abstract void WritePropertyName(string name);
+
+        protected abstract void WriteNullValue();
+
+        protected abstract void WriteNumberValue(double numberValue);
+
+        protected abstract void WriteStringValue(string stringValue);
+
+        protected abstract void WriteBooleanValue(bool booleanValue);
+
+        protected abstract void WriteDateTimeValue(DateTime dateTimeValue);
+
+        internal static readonly DType DType_Table = new (DKind.Table);
+        internal static readonly DType DType_Record = new (DKind.Record);
+
+        internal string Serialize(OpenApiSchema schema, IEnumerable<NamedValue> fields)
+        {
+            StartSerialization(schema);
+            WriteObject(null, schema, fields);           
+            return GetResult();
+        }
+
+        private void WriteObject(string objectName, OpenApiSchema schema, IEnumerable<NamedValue> fields)
+        {
+            StartObject(objectName);
+
+            foreach (var property in schema.Properties)
+            {
+                var namedValue = fields.FirstOrDefault(nv => nv.Name.Equals(property.Key, StringComparison.OrdinalIgnoreCase));
+
+                if (namedValue == null)
+                {
+                    throw new NotImplementedException($"Missing property {property.Key}, object is too complex or not supported");
+                }
+
+                WriteProperty(property.Key, property.Value, namedValue.Value);
```

Comment:
WritePropertyName should be here (lifted up, outside the recursion). 

[https://github.com/microsoft/PowerApps-Language-Tooling/blob/master/src/PAModel/Utility/JsonNormalizer.cs](https://github.com/microsoft/PowerApps-Language-Tooling/blob/master/src/PAModel/Utility/JsonNormalizer.cs)

Reply:
this would also allow us to handle top-level cases that aren't objects.